### DISCUSSION TITLE
ZFS data corruption bug

### DISCUSSION THREAD

**Discussion by silenius** [zrepl/zrepl#774] [Q&A]
Hello,

I am using exclusively zrepl to backup our (FreeBSD) ZFS-based servers, works fine. After years of usage and readings I was convinced that "ZFS can never fails". After https://github.com/openzfs/zfs/issues/15526 it looks like I was a little bit too optimistic, although I'm still a big fan of ZFS. Fortunately we didn't has any corruption but it would have been a disaster if that had been the case ... (even more with backups being corrupted too).
I was wondering if zrepl could include some kind of "safety net" as a workaround, by maintaining a list of checksum of backuped files (for example) so that a comparaison could easily be done at a certain point in time?

Thanks!

**Comment by InsanePrawn** [comment]
to me, that sounds way out of scope for zrepl.

you could probably hook something up via zrepl's hook system though?

**Comment by 0x3333** [comment]
`zrepl` is not a backup solution, is a snapshot one. You should have a proper backup solution that tests your backups. Nothing is safe in this world. 

I use `Kopia` as a backup solution for local and remote copies, it has checksum and all the good stuff a backup needs to have.

**Comment by silenius** [comment]
> `zrepl` is not a backup solution, is a snapshot one. You should have a proper backup solution that tests your backups. Nothing is safe in this world.
> 
> I use `Kopia` as a backup solution for local and remote copies, it has checksum and all the good stuff a backup needs to have.

I don't follow you... why don't you consider zrepl as a "backup solution"? The word "backup" is mentionned several times in the manual and I don't see how it couldn't be used a a backup solution. I was using Bacula before and it took several hours (sometimes more than a day) to perform an incremental backup (checking file one by one isn't scalable)

**Comment by 0x3333** [comment]
You're right, what I meant is that it is not the whole picture... The snapshot is not the backup itself, you need more tooling to have everything sorted out.

Anyway, I don't believe that adding checksums for example would be even feasible.

**Comment by silenius** [comment]
The only missing thing is a searchable catalog-like, otherwise I think you can achieve everything with zfs (send/recv, clone the snapshot, promote it if you want, etc).

**Comment by 0x3333** [comment]
It's important to check the snapshots, I don't like relying only on one tool for consistency.

**Comment by brenc** [comment]
This is why I maintain a remote server running Debian Linux (primary servers are FreeBSD 14) and rsync data over from the primaries. Then I do local snapshots. There are a number of other options. rclone direct to B2 is pretty easy to get set up and is fairly inexpensive, for example.

**Comment by problame** [comment]
zrepl's scope is ZFS replication, which can be used as one component of your backup strategy.

Integration with VFS-level backup is out of scope.
